pkg/monitors/cloudfoundry/conversion.go

```diff
@@ -0,0 +1,57 @@
+package cloudfoundry
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"code.cloudfoundry.org/go-loggregator/rpc/loggregator_v2"
+	"github.com/davecgh/go-spew/spew"
+	"github.com/signalfx/golib/v3/datapoint"
+)
+
+func envelopeToDatapoints(env *loggregator_v2.Envelope) ([]*datapoint.Datapoint, error) {
+	// We intentionally modify the Tags map on the envelope, assuming that the
+	// loggregator code that generated it is not going to reuse envelope
+	// instances or tag maps.
+	dims := env.Tags
+
+	if env.SourceId != "" {
+		dims["source_id"] = env.SourceId
+	}
+
+	if env.InstanceId != "" {
+		dims["instance_id"] = env.InstanceId
+	}
+
+	var metricType datapoint.MetricType
+
+	namesToValues := make(map[string]float64)
+
+	switch m := env.Message.(type) {
+	case *loggregator_v2.Envelope_Log:
+	case *loggregator_v2.Envelope_Counter:
+		metricType = datapoint.Counter
+		namesToValues[m.Counter.GetName()] = float64(m.Counter.GetTotal())
+	case *loggregator_v2.Envelope_Gauge:
+		metricType = datapoint.Gauge
+		for name, gauge := range m.Gauge.GetMetrics() {
+			namesToValues[name] = gauge.Value
+		}
+	case *loggregator_v2.Envelope_Timer:
+	case *loggregator_v2.Envelope_Event:
+	default:
+		return nil, fmt.Errorf("cannot convert envelope %v to SignalFx datapoints", spew.Sdump(env))
+	}
+
+	var dps []*datapoint.Datapoint
+	for name, val := range namesToValues {
+		dps = append(dps, datapoint.New(cleanupName(name), dims, datapoint.NewFloatValue(val), metricType, time.Unix(0, env.Timestamp)))
```

Comment:
What do you think about adding a prefix name to all metrics (something like `pcf_firehose`) to make it easier to search?

Reply:
The metrics will be prefixed with the name of the source/origin within CloudFoundry that emitted them.  We generally should avoid metric prefixes that are something really generic like the platform or mechanism by which they were derived (e.g. the firehose).

If you want to make them easy to search, you could add a global dimension in your agent config that gave all of them a common fixed dimension to search on.